src/org/oidc/service/base/ServiceContext.java

```diff
@@ -0,0 +1,328 @@
+                keyJar.addKeyBundle(ISSUER,keyBundle);
+            }
+        }
+    }
+
+    /**
+     A 1<->1 map is maintained between a URL pointing to a file and
+     the name of the file in the file system.
+
+     As an example if the base_url is 'https://example.com' and a jwks_uri
+     is 'https://example.com/jwks_uri.json' then the filename of the
+     corresponding file on the local filesystem would be 'jwks_uri'.
+     Relative to the directory from which the RP instance is run.
+     * @param webName the published URL
+     * @return local filename
+     **/
+    public String fileNameFromWebname(String webName) throws ValueException {
+        if(Strings.isNullOrEmpty(webName)) {
+            throw new IllegalArgumentException("null or empty webName");
+        }
+
+        if(!webName.startsWith(this.baseUrl)) {
+            throw new ValueException("Webname does not match baseUrl");
+        }
+
+        webName = webName.substring(this.baseUrl.length());
+        if(webName.startsWith("/")) {
+            return webName.substring(1);
+        } else {
+            return webName;
+        }
+    }
+
+    /**
+     Need to generate a redirectUri path that is unique for an OP/RP combo.  This is to counter the mix-up attack.
+
+     * @param requestsDirectory the leading path
+     * @return a list of one unique URL
+     **/
+    public List<String> generateRequestUris(String requestsDirectory) throws NoSuchAlgorithmException {
+        MessageDigest messageDigest = MessageDigest.getInstance(SHA_256);
+        if(this.providerConfigurationResponse.getClaims() != null
+                && this.providerConfigurationResponse.getClaims().get(ClaimType.ISSUER) != null
+                && this.providerConfigurationResponse.getClaims().get(ClaimType.ISSUER) instanceof List
+                && !((List) this.providerConfigurationResponse.getClaims().get(ClaimType.ISSUER)).isEmpty()) {
+            for (String issuer : ((List<String>) this.providerConfigurationResponse.getClaims().get(ClaimType.ISSUER))) {
+                messageDigest.update(issuer.getBytes());
+            }
+        } else {
+            messageDigest.digest(this.issuer.getBytes());
+        }
+        messageDigest.digest(this.baseUrl.getBytes());
+        if(!requestsDirectory.startsWith("/")) {
+            return Arrays.asList(this.baseUrl + "/" + requestsDirectory + "/" + messageDigest.digest());
+        } else {
+            return Arrays.asList(this.baseUrl + requestsDirectory + "/" + messageDigest.digest());
+        }
+    }
+
+    public KeyJar getKeyJar() {
+        return keyJar;
+    }
+
+    public void setKeyJar(KeyJar keyJar) {
+        this.keyJar = keyJar;
+    }
+
+    public ServiceContextConfig getConfig() {
+        return config;
+    }
+
+    public void setConfig(ServiceContextConfig config) {
+        this.config = config;
+    }
+
+    public ProviderConfigurationResponse getProviderConfigurationResponse() {
```

Comment:
@rohe is it correct assumption that ProviderConfigurationResponse and RegistrationResponse are independent of services? That's why it is okay to use these responses in ServiceContext.

Reply:
There is only one service each that can modify their content. But there are several that needs to access the information in them.